shell/common/pointer_data_dispatcher.h

```diff
@@ -0,0 +1,172 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef POINTER_DATA_DISPATCHER_H_
+#define POINTER_DATA_DISPATCHER_H_
+
+#include "flutter/runtime/runtime_controller.h"
+#include "flutter/shell/common/animator.h"
+
+namespace flutter {
+
+class PointerDataDispatcher;
+
+//------------------------------------------------------------------------------
+/// The `Engine` pointer data dispatcher that forwards the packet received from
+/// `PlatformView::DispatchPointerDataPacket` on the platform thread, to
+/// `Window::DispatchPointerDataPacket` on the UI thread.
+///
+/// This class is used to filter the packets so the Flutter framework on the UI
+/// thread will receive packets with some desired properties. See
+/// `SmoothPointerDataDispatcher` for an example which filters irregularly
+/// delivered packets, and dispatches them in sync with the VSYNC signal.
+///
+/// This object will be owned by the engine because it relies on the engine's
+/// `Animator` (which owns `VsyncWaiter`) and `RuntomeController` to do the
+/// filtering. This object is currently designed to be only called from the UI
+/// thread (no thread safety is guaranteed).
+///
+/// The `PlatformView` decides which subclass of `PointerDataDispatcher` is
+/// constructed by sending a `PointerDataDispatcherMaker` to the engine's
+/// constructor in `Shell::CreateShellOnPlatformThread`. This is needed because:
+///   (1) Different platforms (e.g., Android, iOS) have different dispatchers
+///       so the decision has to be made per `PlatformView`.
+///   (2) The `PlatformView` can only be accessed from the PlatformThread while
+///       this class (as owned by engine) can only be accessed in the UI thread.
+///       Hence `PlatformView` creates a `PointerDataDispatchMaker` on the
+///       platform thread, and send it to the UI thread for the final
+///       construction of the `PointerDataDispatcher`.
+class PointerDataDispatcher {
+ public:
+  /// The interface for Engine to implement.
+  class Delegate {
+   public:
+    /// Actually dispatch the packet using Engine's `animator_` and
+    /// `runtime_controller_`.
+    virtual void DoDispatchPacket(std::unique_ptr<PointerDataPacket> packet,
+                                  uint64_t trace_flow_id) = 0;
+
+    //----------------------------------------------------------------------------
+    /// @brief    Schedule a secondary callback to be executed right after the
+    ///           main `Animator::RequestFrame` callback.
+    ///
+    ///           If there is no RequestFrame callback, this secondary callback
+    ///           will still be executed at vsync.
+    ///
+    ///           This callback is used to provide the vsync signal needed by
+    ///           `SmoothPointerDataDispatcher`.
+    virtual void ScheduleSecondaryVsyncCallback(
```

Comment:
I wonder if "secondary" is the best name for this.  Unity has 2 different callbacks "Update" which is tied to rendering a frame and "FixedUpdate" which is called at a regular interval.  Does ScheduleFixedVsyncCallback sound better?